src/coreclr/System.Private.CoreLib/src/System/String.CoreCLR.cs

```diff
@@ -11,30 +12,25 @@ public partial class String
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern string FastAllocateString(int length);
 
-        // Set extra byte for odd-sized strings that came from interop as BSTR.
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal extern void SetTrailByte(byte data);
-        // Try to retrieve the extra byte - returns false if not present.
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal extern bool TryGetTrailByte(out byte data);
-
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private extern string Intern();
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private extern string? IsInterned();
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "String_Intern")]
+        private static partial void Intern(StringHandleOnStack src);
 
         public static string Intern(string str)
         {
             ArgumentNullException.ThrowIfNull(str);
-
-            return str.Intern();
+            Intern(new StringHandleOnStack(ref str!));
```

Comment:
Is there any prospect of Roslyn being able to infer that after the check above, str is not null?

Reply:
The problem is that `StringHandleOnStack` constructor takes `ref string?`. It would be ok for the constructor to set the string to null. Roslyn is correctly complaining about that without `!`.

To solve this without `!`, it would need to be written like this:
```
string? str2 = str;
Intern(new StringHandleOnStack(ref str2));
```
Or we would need to create multiple overloads of StringHandleOnStack with different nullability annotations. Not worth it. StringHandleOnStack is complicated enough as is.